File: guests/haskell/grow.hs
{-# LANGUAGE OverloadedStrings #-}

{- The grow conformance scene, Haskell port — see guests/rust/grow.rs.
   Every child of the column and of the row is a grower, so each split
   is exactly weight/Σweight: 1,2,1 divide the column 25/50/25 and the
   row's 1,3 divide its width 25/75.

   The textarea's handler is a no-op because 'textareaOn' is the only
   spelling this binding has. -}

import qualified Data.Text as T
import KayaApp

main :: IO ()
main = kayaMain $ \app -> do
  buildTx app $ do
    probe <- signal (T.pack "grow probe")
    one <- signal (T.pack "one")

    root <-
      column
        [ labelBound probe [Grow 1], -- label#0
          textareaOn (\_ -> return ()) [Grow 2], -- textarea#0
          row
            [Grow 1, Spacing 12]
            [ labelBound one [Grow 1], -- label#1
              buttonOn "three" (return ()) [Grow 3]
            ]
        ]
    mount root

File: guests/rust/grow.rs
//! The grow conformance scene (tools/scenes/grow.steps): EVERY CHILD IS A
//! GROWER, and there is exactly ONE column and ONE row.

pub(crate) fn app(ctx: kaya::AppCtx) {
    let msgs = kaya::Messages::<()>::new();
    ctx.apply(|tx| {
        let probe = tx.signal("grow probe");
        let one = tx.signal("one");

        let root = tx.column(|tx| {
            // THE SHARE ARITHMETIC (docs/traps.md): recompute it whenever a
            // weight moves, or the scene measures a platform minimum.
            tx.label(probe).grow(1.0); // label#0
            // Only expect_fills beside expect_shares tells "took its track"
            // from "was given one".
            tx.textarea().grow(2.0); // textarea#0
            // A non-default gap: a backend that ignores the write fails.
            tx.row(|tx| {
                tx.label(one).grow(1.0); // label#1
                tx.button("three").grow(3.0);
            })
            .grow(1.0)
            .spacing(12.0);
        })
        .id();
        tx.mount(root);
    });

    // Blocks until the harness sends Shutdown.
    while msgs.next(&ctx).is_some() {}
}

fn main() {
    kaya::run(app)
}
